ffmpeg/ffmpeg.go

```diff
@@ -107,41 +108,163 @@ type TranscodeResults struct {
+)
+
+type ColorDepthBits int
+
+const (
+	ColorDepth8Bit  ColorDepthBits = 8
+	ColorDepth10Bit ColorDepthBits = 10
+	ColorDepth12Bit ColorDepthBits = 12
+	ColorDepth16Bit ColorDepthBits = 16
+)
+
+type ChromaSubsampling int
+
+const (
+	ChromaSubsampling420 ChromaSubsampling = iota
+	ChromaSubsampling422
+	ChromaSubsampling444
+)
+
+func (self PixelFormat) Properties() (ChromaSubsampling, ColorDepthBits, error) {
+	switch self.RawValue {
+	case C.AV_PIX_FMT_YUV420P:
+		return ChromaSubsampling420, 8, nil
+	case C.AV_PIX_FMT_YUYV422:
+		return ChromaSubsampling422, 8, nil
+	case C.AV_PIX_FMT_YUV422P:
+		return ChromaSubsampling422, 8, nil
+	case C.AV_PIX_FMT_YUV444P:
+		return ChromaSubsampling444, 8, nil
+	case C.AV_PIX_FMT_UYVY422:
+		return ChromaSubsampling422, 8, nil
+	case C.AV_PIX_FMT_NV12:
+		return ChromaSubsampling420, 8, nil
+	case C.AV_PIX_FMT_NV21:
+		return ChromaSubsampling420, 8, nil
+	case C.AV_PIX_FMT_YUV420P10BE:
+		return ChromaSubsampling420, 10, nil
+	case C.AV_PIX_FMT_YUV420P10LE:
+		return ChromaSubsampling420, 10, nil
+	case C.AV_PIX_FMT_YUV422P10BE:
+		return ChromaSubsampling422, 10, nil
+	case C.AV_PIX_FMT_YUV422P10LE:
+		return ChromaSubsampling422, 10, nil
+	case C.AV_PIX_FMT_YUV444P10BE:
+		return ChromaSubsampling444, 10, nil
+	case C.AV_PIX_FMT_YUV444P10LE:
+		return ChromaSubsampling444, 10, nil
+	case C.AV_PIX_FMT_YUV420P16LE:
+		return ChromaSubsampling420, 16, nil
+	case C.AV_PIX_FMT_YUV420P16BE:
+		return ChromaSubsampling420, 16, nil
+	case C.AV_PIX_FMT_YUV422P16LE:
+		return ChromaSubsampling422, 16, nil
+	case C.AV_PIX_FMT_YUV422P16BE:
+		return ChromaSubsampling422, 16, nil
+	case C.AV_PIX_FMT_YUV444P16LE:
+		return ChromaSubsampling444, 16, nil
+	case C.AV_PIX_FMT_YUV444P16BE:
+		return ChromaSubsampling444, 16, nil
+	case C.AV_PIX_FMT_YUV420P12BE:
+		return ChromaSubsampling420, 12, nil
+	case C.AV_PIX_FMT_YUV420P12LE:
+		return ChromaSubsampling420, 12, nil
+	case C.AV_PIX_FMT_YUV422P12BE:
+		return ChromaSubsampling422, 12, nil
+	case C.AV_PIX_FMT_YUV422P12LE:
+		return ChromaSubsampling422, 12, nil
+	case C.AV_PIX_FMT_YUV444P12BE:
+		return ChromaSubsampling444, 12, nil
+	case C.AV_PIX_FMT_YUV444P12LE:
+		return ChromaSubsampling444, 12, nil
+	default:
+		return 0, 0, ErrTranscoderPixelformat
+	}
+}
+
+type CodecStatus int
+
+const (
+	CodecStatusInternalError CodecStatus = -1
+	CodecStatusOk            CodecStatus = 0
+	CodecStatusNeedsBypass   CodecStatus = 1
+	CodecStatusMissing       CodecStatus = 2
+)
+
+func GetCodecInfo(fname string) (CodecStatus, string, string, PixelFormat, error) {
 	var acodec, vcodec string
 	cfname := C.CString(fname)
 	defer C.free(unsafe.Pointer(cfname))
 	acodec_c := C.CString(strings.Repeat("0", 255))
 	vcodec_c := C.CString(strings.Repeat("0", 255))
 	defer C.free(unsafe.Pointer(acodec_c))
 	defer C.free(unsafe.Pointer(vcodec_c))
-	bres := int(C.lpms_get_codec_info(cfname, vcodec_c, acodec_c))
+	var params_c C.codec_info
```

Comment:
Should we free `params_c` struct too? Also I think eventually we'd want to return the struct from this function, not just use it internally in C interop.

Reply:
Good question for deleting `params_c` struct. So far i have found the answer is: no explicit free on struct object created by go code.

Using IDE i can see the definition as:
```
type _Ctype_struct_s_codec_info struct {
	video_codec	*_Ctype_char
	audio_codec	*_Ctype_char
	pixel_format	_Ctype_int
	_		[4]byte
}
```
in (one can say generated file) `./build/0b/0bcf360880d821f4528cb9ca62dac5d996f2f71206d9362686c18ac37cae4b1f-d`

In this case go-created `params_c` struct is passed to C code to be populated. There are many constraints on how C code is allowed to manipulate given parameter and i think we are safe to do this.

Please correct me if i am wrong, `params_c` should be collected by go and allocated strings with `C.CString()` we free manually.

[Relevant research reference](https://stackoverflow.com/questions/10866195/stack-vs-heap-allocation-of-structs-in-go-and-how-they-relate-to-garbage-collec)

Reply:
Interesting! So we don't need to free struct explicitly in this case, because it's not explicitly allocated, this makes sense.